src/curl.c

```diff
@@ -363,6 +364,62 @@ int swupd_curl_get_file(const char *url, char *filename, struct file *file,
 	return err;
 }
 
+void swupd_curl_test_resume(void)
+{
+#define RESUME_BYTE_RANGE 2
+
+	CURLcode curl_ret;
+	char *version_string = NULL;
+	char *url = NULL;
+
+	if (!curl) {
+		abort();
+	}
+
+	curl_easy_reset(curl);
+	version_string = malloc(LINE_MAX);
+	if (version_string == NULL) {
+		abort();
+	}
+
+	string_or_die(&url, "%s/version/format%s/latest", version_url, format_string);
+
+	curl_ret = curl_easy_setopt(curl, CURLOPT_RESUME_FROM_LARGE, (curl_off_t)RESUME_BYTE_RANGE);
+	if (curl_ret != CURLE_OK) {
+		goto exit;
+	}
+
+	curl_ret = curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, swupd_download_version_to_memory);
+	if (curl_ret != CURLE_OK) {
+		goto exit;
+	}
+	curl_ret = curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)version_string);
```

Comment:
Ok so I went ahead and did little investigation, according to RFC 7233
snipet:
<-----snipet---->

An origin server that supports byte-range requests for a given target
   resource MAY send

     Accept-Ranges: bytes

   to indicate what range units are supported.  A client MAY generate
   range requests without having received this header field for the
   resource involved.  Range units are defined in Section 2
<https://tools.ietf.org/html/rfc7233#section-2>.

   A server that does not support any kind of range request for the
   target resource MAY send

     Accept-Ranges: none

   to advise the client not to attempt a range request.


</---snipet----/>

This does not have a "must" advertise if Accept-Ranges: none in case no
Ranges is accepted, having said that I went ahead and inspected headers in
one server that advertises Accept-Ranges and other that doesn't:


   1. Accept-Ranges:
   bytes
   2. Connection:
   Keep-Alive
   3. Content-Length:
   5
   4. Date:
   Wed, 04 May 2016 17:24:35 GMT
   5. ETag:
   "5-531e5c90e9080"
   6. Keep-Alive:
   timeout=5, max=100
   7. Last-Modified:
   Tue, 03 May 2016 01:11:14 GMT
   8. Server:
   Apache/2.4.6 (CentOS)



   1. Content-Length:
   268
   2. Content-type:
   text/html; charset=ANSI_X3.4-1968
   3. Date:
   Wed, 04 May 2016 17:19:10 GMT
   4. Server:
   SimpleHTTP/0.6 Python/2.7.11

As you can notice, SimpleHTTPServer don't even advertise Accept-Ranges
support, which means basically more validation to check one for advertising
Accept-Ranges: none and another for not even having that header in place.

Having said that, the already proposed patch is cheap because only is
requesting a couple of bytes and we get an straight answer code which we
can validate pretty simple, also if I request only the headers when
CURLOPT_RESUME_FROM_LARGE is set, somehow is skipped and not returning the
expected error.

Conclusion, IMHO this patch is simpler as going to examine the headers
which will require to create a linked list and parsing/checking for
different cases.


On Tue, May 3, 2016 at 1:23 PM, Patrick McCarty <notifications@github.com>
wrote:

> In src/curl.c
> <https://github.com/clearlinux/swupd-client/pull/59#discussion_r61930258>:
>
> > +	if (version_string == NULL) {
> > +		abort();
> > +	}
> > +
> > +	string_or_die(&url, "%s/version/format%s/latest", version_url, format_string);
> > +
> > +	curl_ret = curl_easy_setopt(curl, CURLOPT_RESUME_FROM_LARGE, (curl_off_t)RESUME_BYTE_RANGE);
> > +	if (curl_ret != CURLE_OK) {
> > +		goto exit;
> > +	}
> > +
> > +	curl_ret = curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, swupd_download_version_to_memory);
> > +	if (curl_ret != CURLE_OK) {
> > +		goto exit;
> > +	}
> > +	curl_ret = curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)version_string);
>
> I think it would worth testing downloading the header only, and see if the
> CURLE_RANGE_ERROR error is raised for that case.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/clearlinux/swupd-client/pull/59/files/eecdce9de214c06e806037ec1cf24e17f8b8e7e4#r61930258>
>